protocols/direct-funding.go

```diff
@@ -0,0 +1,350 @@
+			sum := big.NewInt(0)
+			threshold := big.NewInt(0)
+			myShare := big.NewInt(0)
+
+			for i, allocation := range assetExit.Allocations {
+				sum = sum.Add(sum, allocation.Amount)
+
+				if i < int(init.MyIndex) {
+					threshold = threshold.Add(threshold, allocation.Amount)
+				} else if i == int(init.MyIndex) {
+					myShare = myShare.Add(myShare, allocation.Amount)
+				}
+			}
+
+			init.FullyFundedThreshold[assetAddress] = sum
+			init.MyDepositSafetyThreshold[assetAddress] = threshold
+			init.MyDepositTarget[assetAddress] = myShare.Add(myShare, threshold)
+		}
+	}
+
+	init.PostFundSigned = make([]bool, len(initialState.Participants))
+	init.OnChainHolding = types.Funds{}
+
+	return init, nil
+}
+
+// Public methods on the DirectFundingObjectiveState
+
+func (s DirectFundingObjectiveState) Id() ObjectiveId {
+	return ObjectiveId("DirectFunding-" + s.ChannelId.String())
+}
+
+func (s DirectFundingObjectiveState) Approve() Objective {
+	updated := s.clone()
+	// todo: consider case of s.Status == Rejected
+	updated.Status = Approved
+
+	return updated
+}
+
+func (s DirectFundingObjectiveState) Reject() Objective {
+	updated := s.clone()
+	updated.Status = Rejected
+	return updated
+}
+
+// Update receives an ObjectiveEvent, applies all applicable event data to the DirectFundingObjectiveState,
+// and returns the updated state
+func (s DirectFundingObjectiveState) Update(event ObjectiveEvent) (Objective, error) {
+	if s.ChannelId != event.ChannelId {
+		return s, errors.New("event and objective channelIds do not match")
+	}
+
+	updated := s.clone()
+
+	for _, sig := range event.Sigs {
+		var turnNum int
+		if updated.prefundComplete() {
+			turnNum = 1
+		} else {
+			turnNum = 0
+		}
+		updated, _ = updated.signatureRecieved(sig, turnNum)
+	}
+
+	if event.Holdings != nil {
+		updated.OnChainHolding = event.Holdings
+	}
+
+	return updated, nil
+}
+
+// Crank inspects the extended state and declares a list of Effects to be executed
+// It's like a state machine transition function where the finite / enumerable state is returned (computed from the extended state)
+// rather than being independent of the extended state; and where there is only one type of event ("the crank") with no data on it at all
+func (s DirectFundingObjectiveState) Crank(secretKey *[]byte) (Objective, SideEffects, WaitingFor, error) {
+	updated := s.clone()
+
+	// Input validation
+	if updated.Status != Approved {
+		return updated, NoSideEffects, WaitingForNothing, ErrNotApproved
+	}
+
+	// Prefunding
+	if !updated.PreFundSigned[updated.MyIndex] {
+		// todo: {SignPreFundEffect(updated.ChannelId)} as SideEffects{}
+		return updated, NoSideEffects, WaitingForCompletePrefund, nil
+	}
+
+	if !updated.prefundComplete() {
+		return updated, NoSideEffects, WaitingForCompletePrefund, nil
+	}
+
+	// Funding
+	fundingComplete := updated.fundingComplete(updated.OnChainHolding) // note all information stored in state (since there are no real events)
```

Comment:
There's some unnecessary duplication, here. We don't need to pass in a property of `updated`, since `.updated.fundingComplete` can simply read it from the receiver?